tests/Unit/Domain/CoordinateMaps/Test_CylindricalEndcap.cpp

```diff
@@ -0,0 +1,224 @@
+             "is not inside sphere_two");
+      // Should be two real roots. Choose the positive one.
+      const double r_max = positive_root(a, b, c);
+
+      const double proj_radius = unit_dis(gen) * r_max;
+      return std::array<double, 3>{
+          center_one[0] + proj_radius * sin(beta) * cos(phi),
+          center_one[1] + proj_radius * sin(beta) * sin(phi),
+          max_proj_center_z - proj_radius * cos(beta)};
+    } else {
+      // The more difficult case.
+
+      // We may need to try several values of alpha.
+      // When one works, exit.
+      // Failure is rare; most of the time it should succeed on the first try.
+      while (true) {
+        // Choose some smaller cone inside of cone_regular.
+        // We will place proj_center on this smaller cone.
+        // We do not allow the smaller cone to go all the way to
+        // cone_regular_opening_angle.
+        const double alpha = unit_dis(gen) * 0.95 * cone_regular_opening_angle;
+
+        // Consider the circle at which the smaller cone intersects
+        // cone_opening, and find the distance from cone_regular_apex to
+        // this circle.
+        const double radius_coord_circle =
+            (cone_regular_apex[2] - max_proj_center_z) *
+            tan(max_opening_angle) / (tan(max_opening_angle) - tan(alpha)) /
+            cos(alpha);
+
+        // Choose an azimuthal coordinate for the point on the cone.
+        const double phi = angle_dis(gen);
+
+        // Choose a radius for the point on the cone that will be
+        // proj_center. The point should be inside sphere_two.  So
+        // determine the radius at which the point intersects
+        // sphere_two. This radius is the solution of a quadratic equation
+        // ax^2 + bx + c = 0
+        const double a = 1.0;
+        const double b =
+            2.0 * (-(cone_regular_apex[2] - center_two[2]) * cos(alpha) +
+                   (center_one[0] - center_two[0]) * sin(alpha) * cos(phi) +
+                   (center_one[1] - center_two[1]) * sin(alpha) * sin(phi));
+        const double c = square(cone_regular_apex[2] - center_two[2]) +
+                         square(center_one[0] - center_two[0]) +
+                         square(center_one[1] - center_two[1]) -
+                         square(radius_two);
+
+        double x0 = std::numeric_limits<double>::signaling_NaN();
+        double x1 = std::numeric_limits<double>::signaling_NaN();
+        const int num_real_roots = gsl_poly_solve_quadratic(a, b, c, &x0, &x1);
+        double r_max = 0.0;
+        if (num_real_roots == 2) {
+          // Take the largest root.  The smallest one is negative if
+          // cone_regular_apex[0] is inside sphere_two, positive if
+          // cone_regular_apex[0] is outside sphere_two.
+          r_max = std::max(x0, x1);
+        } else if (num_real_roots == 1) {
+          r_max = x0;
+        } else {
+          ASSERT(false, "No roots were found. Something is horribly wrong.");
```

Comment:
any chance I could convince you to give some hint (in the ASSERT message) as to what is "horribly wrong" and where I should look to fix it? :D